File: cpp/ros2_pubsub_demo/sub.cc
#include <iostream>
#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/string.hpp>

void callback(const std_msgs::msg::String::SharedPtr msg) {
    std::cout << "sub " << msg->data.c_str() << std::endl;
}


int main(int argc, char **argv) {
    std::cout << "Hello, ROS2 Sub!" << std::endl;
    rclcpp::init(argc, argv);
    auto node = rclcpp::Node::make_shared("sub_node");
    auto sub = node->create_subscription<std_msgs::msg::String>("/pub", 1, callback);
    rclcpp::spin(node);
    return 0;
}

File: py/ros2_pubsub_demo/sub.py
import rclpy
from rclpy.node import Node

from std_msgs.msg import String


class MinimalSubscriber(Node):

    def __init__(self):
        super().__init__('minimal_subscriber')
        self.subscription = self.create_subscription(
            String,
            'topic',
            self.listener_callback,
            10)
        self.subscription  # prevent unused variable warning

    def listener_callback(self, msg):
        print("calback:", msg)


if __name__ == "__main__":
    
    print("ros2 py sub")

    rclpy.init()

    node = MinimalSubscriber()

    rclpy.spin(node)

    node.destroy_node()
    rclpy.shutdown()


